### ISSUE TITLE
Characters clipped in Chrome when rendering PDFs that use variable weight fonts

### ISSUE THREAD

**Issue by mikecoulhs** [mozilla/pdf.js#14941]
Indesign was used to create the following PDFs.
- [indesign-test-1.pdf](https://github.com/mozilla/pdf.js/files/8732551/indesign-test-1.pdf)
- [indesign-test-2.pdf](https://github.com/mozilla/pdf.js/files/8732557/indesign-test-2.pdf)

The document uses the [RedHatDisplay fonts from Google](https://fonts.google.com/specimen/Red+Hat+Display).  Google provides two versions of the fonts, "variable weight" and "static".  The first document was created with the "variable weight" font installed.  The second document was created with the "static" fonts installed.

The first one renders with the individual characters incorrectly clipped, while the second one renders correctly.

Configuration:
- Web browser and its version: Chrome Version 100.0.4896.127 (Official Build) (x86_64)
- Operating system and its version: MacOS Catalina 10.15.7
- PDF.js version: Version at https://mozilla.github.io/pdf.js/web/viewer.html
- Is a browser extension: No

Steps to reproduce the problem:
1. Using the pdfjs online viewer, open both PDFs in chrome and compare.  Notice the clipped characters in the first image.

indesign-test1.pdf:
![image](https://user-images.githubusercontent.com/22301903/169365214-13533e13-4047-4be8-b7fb-f3fc4d868bdd.png)

indesign-test2.pdf:
![image](https://user-images.githubusercontent.com/22301903/169365255-e45f75cf-6956-4b70-a18b-93396eef3d51.png)

The problem also appears to occur in Firefox 100.0.1 (64-bit)
The problem does NOT appear to occur in Safari (Version 15.0 (15612.1.29.41.4, 15612))

**Comment by calixteman** [comment]
The bug has already been reported in Firefox:
https://bugzilla.mozilla.org/show_bug.cgi?id=1756855

**Comment by jrmuizel** [comment]
I think this problem may also be happening on https://ontarioliberal.ca/wp-content/uploads/2022/05/Ontario-Liberal-Platform.pdf

**Comment by mikecoulhs** [comment]
@calixteman it's been some time, but we continue to encounter this issue with different documents.  I notice in the link you said that you were unable to repro with the linked files.  It appears that this only happens in Chrome, and I am still able to see the issue in with those files in Chrome Version 139.0.7258.155 (Official Build) (arm64) with the version 5.3.31 of the PDF.js viewer as of today.

<img width="687" height="347" alt="Image" src="https://github.com/user-attachments/assets/aef03795-8db8-4565-a096-3db911b1cec3" />

If you could take another look at this issue, I would appreciate it.  Thanks!

**Comment by mikecoulhs** [comment]
@calixteman checking in again in the hopes that this might get on you radar.  Thanks!

**Comment by calixteman** [comment]
In looking at the output of `ttx`, the bounding box of the glyphs look wrong and it's maybe why they're clipped.
I can't reproduce the bug on Mac with Firefox but I still can reproduce it with Chrome.
@drott maybe this a bug for you, wdyt ?
Here's the font created by pdf.js:
[SkQbvle5.ttf.zip](https://github.com/user-attachments/files/23269813/SkQbvle5.ttf.zip)

**Comment by drott** [comment]
Indeed, in `SkQbvle5.ttf` the TrueType contour `i` glyph is clipped, similarly as in the artifacts above:

<img width="255" height="428" alt="Image" src="https://github.com/user-attachments/assets/9c3fc866-50f6-4b1b-a9b2-7a591bc1bd87" />

TTX `i` `glyf` table entry:

```
    <TTGlyph name="i" xMin="52" yMin="0" xMax="128" yMax="679">
      <contour>
        <pt x="50" y="0" on="1"/>
        <pt x="50" y="529" on="1"/>
        <pt x="212" y="529" on="1"/>
        <pt x="212" y="0" on="1"/>
      </contour>
      <contour>
        <pt x="131" y="578" on="1"/>
        <pt x="94" y="578" on="0"/>
        <pt x="42" y="630" on="0"/>
        <pt x="42" y="667" on="1"/>
        <pt x="42" y="704" on="0"/>
        <pt x="94" y="756" on="0"/>
        <pt x="131" y="756" on="1"/>
        <pt x="168" y="756" on="0"/>
        <pt x="220" y="704" on="0"/>
        <pt x="220" y="667" on="1"/>
        <pt x="220" y="630" on="0"/>
        <pt x="168" y="578" on="0"/>
      </contour>
      <instructions/>
    </TTGlyph>
```

From an SVG approximation of the `glyf` table TTX info:

```
<svg width="255" height="255" viewBox="0 0 255 255" xmlns="http://www.w3.org/2000/svg">
  <g transform="translate(0, 255) scale(0.3373, -0.3373)">
    <path 
      fill="black" 
      stroke="black"
      d="
        M 50 0 L 50 529 L 212 529 L 212 0 Z 
        
        M 131 578 
        C 94 578, 42 630, 42 667 
        C 42 704, 94 756, 131 756 
        C 168 756, 220 704, 220 667 
        C 220 630, 168 578, 131 578 
        Z
      "/>
    <rect 
      x="52" 
      y="0" 
      width="76" 
      height="679" 
      stroke="red" 
      fill="none" 
      stroke-width="10"
      stroke-dasharray="20 10" 
      opacity="0.7"/>
  </g>
</svg>
```

> In looking at the output of ttx, the bounding box of the glyphs look wrong and it's maybe why they're clipped.

@calixteman did you mean to imply the original variable font bounding boxes for _RedHatDisplay_  were incorrect, or did you mean you were looking at the clipped PDF.js output font?

This issue was originally reported in 2022 which precludes this being a Fontations font stack issue in Chrome.  

I am not sure but, this might be reproducible in Chrome only because PDF.js chooses a different generator path for this UA?

**Comment by drott** [comment]
Looking at the "Black" named instance of Red Hat Display in [Samsa](https://www.axis-praxis.org/samsa/), we see that the regular/default instance roughly matches the clipping we see for the i glyph (gray shade default instance, contour and arrows show deltas to Black variant)

<img width="347" height="614" alt="Image" src="https://github.com/user-attachments/assets/281468e1-06b8-4baf-b9e1-339620192732" />

I think the underlying issue is, someone is generating bounding boxes that are stemming from the default instance of the font, instead of applying variations to them.

I am not sure who generates the clips - is Illustrator putting the clips into the PDF, and PDF.js turns them into clips in the font file? I tend to think Chrome is correct in adhering to the bounding box that's in the font file here - and the correct bounding boxes should be provided further up in the stack.

**Comment by calixteman** [comment]
> @calixteman did you mean to imply the original variable font bounding boxes for RedHatDisplay were incorrect, or did you mean you were looking at the clipped PDF.js output font?

As you said the subsetter probably made a bad job here...
Here's the font as it is in the pdf:
[font_85_0.zip](https://github.com/user-attachments/files/23302879/font_85_0.zip)

> This issue was originally reported in 2022 which precludes this being a Fontations font stack issue in Chrome.
It's used to be a bug in Firefox too but it isn't anymore.

I didn't check on a Linux machine, but it works correctly with Firefox on win & mac, it's ok too with Chrome on win, ok with Safarti, it seems to be only an issue with Chrome on mac (Tahoe 26.0.1).
I just updated Chrome (142.0.7444.60) and it's the same.

**Comment by calixteman** [comment]
And it's working correctly with the pdf viewer in Chrome.

**Comment by drott** [comment]
Yes, the font from @calixteman's [comment](https://github.com/mozilla/pdf.js/issues/14941#issuecomment-3480524135) [font_85_0.zip](https://github.com/user-attachments/files/23302879/font_85_0.zip) that is in the PDF (which is the font before the one that pdf.js outputs), already contains the too constrained bounding box:

```
    <TTGlyph name="i" xMin="52" yMin="0" xMax="128" yMax="679">
      <contour>
        <pt x="50" y="0" on="1"/>
        <pt x="50" y="529" on="1"/>
        <pt x="212" y="529" on="1"/>
        <pt x="212" y="0" on="1"/>
      </contour>
      <contour>
        <pt x="131" y="578" on="1"/>
        <pt x="94" y="578" on="0"/>
        <pt x="42" y="630" on="0"/>
        <pt x="42" y="667" on="1"/>
        <pt x="42" y="704" on="0"/>
        <pt x="94" y="756" on="0"/>
        <pt x="131" y="756" on="1"/>
        <pt x="168" y="756" on="0"/>
        <pt x="220" y="704" on="0"/>
        <pt x="220" y="667" on="1"/>
        <pt x="220" y="630" on="0"/>
        <pt x="168" y="578" on="0"/>
      </contour>
      <instructions/>
    </TTGlyph>
```

which means that this is an Illustrator bug where Illustrator's variable font subsetting combines the default instance bounding box with the black variant glyphs.

I suggest to reduce the issue to one or two glyphs, in a simple illustrator file and report the issue to Adobe.

**Comment by mikecoulhs** [comment]
Thank you @drott and @calixteman for the further investigation, I appreciate it.  Knowing that this is a bug in Illustrator is helpful.  Given that it renders correctly in Adobe Reader, I suspect Adobe won't recognize (let alone prioritize) the issue is in their product.  Given that PDF.js correctly renders it correctly on all other platforms and browsers (and other renderers are unaffected  as well), might I request that PDF.js consider implementing a workaround for Mac Chrome?  Thank you!

**Comment by calixteman** [comment]
Since the bug is Chrome on Mac only I'm inclined to think that's a Chrome bug.
@drott is it an issue with Fontations or something else ? I thought Fontations was enabled on both Mac and Windows.
The bug has been fixed in Firefox thanks to [bug 1847959](https://bugzilla.mozilla.org/show_bug.cgi?id=1847959) which corresponds to update of OTS containing this specific patch:
https://github.com/khaledhosny/ots/pull/264
and it seems that Chrome is using OTS from 2022 (https://source.chromium.org/chromium/chromium/src/+/main:third_party/ots/README.chromium;l=3)
So maybe "just" updating OTS would fix this issue (in [bug 1847959](https://bugzilla.mozilla.org/show_bug.cgi?id=1847959), it's worth noticing that isn't a pdf.js bug 😃).
We could sanitize more deeply the fonts on our side but in most of the cases it's useless and it'll decrease the performances for just a corner case so I'm not super excited by this idea.

**Comment by mikecoulhs** [comment]
@drott @calixteman I just wanted to say thanks for the detailed explanation.  It sounds like the best I can hope for, (aside from InDesign fixing the root issue) would be for Chromium to pick up the latest OTS.

**Comment by mikecoulhs** [comment]
FWIW, I've opened a ticket against chromium: https://issues.chromium.org/issues/458442526

**Comment by Snuffleupagus** [comment]
According to https://github.com/mozilla/pdf.js/issues/14941#issuecomment-3487401330 this has been fixed in Firefox and according to https://github.com/mozilla/pdf.js/issues/14941#issuecomment-3514288622 it's been reported to Chrome, hence closing this one as it doesn't seem to be a PDF.js issue.
